### PR TITLE
[PORT] Bee's Dynamic updates

### PR DESCRIPTION
**Pull request by lmenvs**
# About The Pull Request

Ports:
https://github.com/BeeStation/BeeStation-Hornet/pull/7438
https://github.com/BeeStation/BeeStation-Hornet/pull/7586
https://github.com/BeeStation/BeeStation-Hornet/pull/7681
https://github.com/BeeStation/BeeStation-Hornet/pull/7953
https://github.com/BeeStation/BeeStation-Hornet/pull/8152
https://github.com/BeeStation/BeeStation-Hornet/pull/8340
https://github.com/BeeStation/BeeStation-Hornet/pull/8347
https://github.com/BeeStation/BeeStation-Hornet/pull/8405

This is just the rules set port.  Once we're good with this, I'll go ahead and port the antag updates they have.

Additionally, I kept most of our threat rules in, so there may be some tweaking for those as needed.  Mainly this should make Dynamic less likely to hoard threat for midround injections.  

We WILL need community feedback on this.

## Why It's Good For The Game

Makes the rounds more interesting!

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

## Changelog

:cl:  itsmeow, Mothblocks, Y0SH1M4S73R, IndieanaJones, Iamgoofball, wesoda25, Watermelon914, tf-4, Ghommie, JohnFulpWillard, Mooshimi, MrMelbert
balance: Winning dynamic revs no longer adds midround threat. Instead, it'll force a heavy ruleset to spawn, and barring that, will spawn a dangerous random event.
balance: Dynamic will not spawn midround traitors, obsessed, or heretics at lowpop if over 15% of the living crew is antag.
balance: Dynamic's threat will now scale the center of the threat curve down by one threat for each player under 30 players readied at roundstart.
fix: Fixed midrounds not logging properly.
fix: Fixed dynamic simulation capping threat based on actual roundstart players ready instead of simulated players ready.
tweak: Dynamic will now count unreadied players as half a player (round up) for ruleset selection purposes, allowing rulesets to get a more accurate depiction of the roundstart population.
code: Improved dynamic's handling of midround execute() failure, and improved the code.
admin: Added messages for if midrounds fail due to execute() failure, and list how much threat is refunded.
server: Dynamic now has an extra configuration value, max_threat_level. Defaults to 100 threat.
server: Added more dynamic logging
server: Added more logging for dynamic execute() failure.
/:cl:

### CODE REVIEW

**Comment by nednaZ** [comment]
We've already got a good handful of these.
Exists: add: Revenant, Space Pirates, and Obsessed have been added to dynamic mode.
Hivemind is currently disabled as I recall

Both of these have already been done
balance: Midround rolls will now happen more frequently, with their power scaling up over time.
balance: Makes wizard a high impact ruleset.

Already fixed
fix: Fixes the lone operative event hardly ever rolling even if the disk is left on the same place for hours on dynamic.
fix: A lack of candidates for certain antags will no longer runtime.
qol: Ghosts will now see the name of who rolled a midround instead of the new midround's character name.

File: code/modules/mob/mob_lists.dm [hunk 1]
```diff
@@ -121,7 +121,14 @@
 /mob/proc/add_to_current_living_antags()
 	if(!SSticker?.mode)
 		return
-	SSticker.mode.current_players[CURRENT_LIVING_ANTAGS] |= src
+
+	if (length(mind.antag_datums) == 0)
```
Comment (dwasint) [inline code/modules/mob/mob_lists.dm:125]: if !length also works here and is arguably better

**Comment by wraith-54321** [comment]
About the threat roles:
keep the traitor weights lower for now as there were major issues with traitor spam(also why is the roundstart antag cap requirement way higher then heretics)
please just dont do the midround revs changes in this, low pop/threat revs just dont sound fun(being fair though I am biased)
I would recommend bumping midround heretic threat up to 15
I dont think wizards need to be more rare, please raise their weights back up
Midrounds nukies already has problems with failing to roll at 3 required candidates, please keep it there

**Comment by polygoblyn** [comment]
> About the threat roles: keep the traitor weights lower for now as there were major issues with traitor spam(also why is the roundstart antag cap requirement way higher then heretics) please just dont do the midround revs changes in this, low pop/threat revs just dont sound fun(being fair though I am biased) I would recommend bumping midround heretic threat up to 15 I dont think wizards need to be more rare, please raise their weights back up Midrounds nukies already has problems with failing to roll at 3 required candidates, please keep it there

Yeah I fucked that up, I meant to have not actually touched the weights, so I'm fixing them back to what they were.  This is basically supposed to just bring us up to date with Bee so I can start porting the antag updates they have.

**Comment by polygoblyn** [comment]
> We've already got a good handful of these. Exists: add: Revenant, Space Pirates, and Obsessed have been added to dynamic mode. Hivemind is currently disabled as I recall
> 
> Both of these have already been done balance: Midround rolls will now happen more frequently, with their power scaling up over time. balance: Makes wizard a high impact ruleset.
> 
> Already fixed fix: Fixes the lone operative event hardly ever rolling even if the disk is left on the same place for hours on dynamic. fix: A lack of candidates for certain antags will no longer runtime. qol: Ghosts will now see the name of who rolled a midround instead of the new midround's character name.

Thanks, I just threw in the changelogs of the PRs that I ported, removed the items that were redundant

File: code/modules/mob/mob_lists.dm (hunk 1 above)
Comment (polygoblyn) [reply code/modules/mob/mob_lists.dm:125]: Should we change the entire file then? They all have the same convention as this

**Comment by polygoblyn** [comment]
I think I will just port the antags instead of updating this

**Pull request closed without merging**